Automattic/jetpack
# Popover: React 19 compatibility - modifying component to use refs instead of findDOMNode

Fixes https://github.com/Automattic/jetpack/issues/37408

## Proposed changes:

Popover functionality:

* The expected behaviour of the Popover component is that it would not render it's content if `props.isVisible` is not true, or if `props.context` is not a DOM element. If  `props.ignoreContext` is set, the idea here is that clicking inside a DOM element that is set as the ignoreContext will not close the popover.
* There are three components that use Popover: `InfoPopover` which is the primary component I've been testing with (on the Jetpack dashboard), but also `FormToggle`, primarily used in `CompactFormToggle`, itself used in the Privacy class, used on the Privacy settings page - `/wp-admin/admin.php?page=jetpack#/privacy`. And lastly `Tooltip`, primarily used on the at-a-glance stats chart, also on the Jetpack dashboard.

The changed code:

* In the old code, `this.domContext` was assigned either a DOM element directly or the result of `ReactDom.findDOMNode`. The `findDOMNode` method is now considered unsafe and deprecated in strict mode.
* `this.domContextRef` and `this.domContainerRef` are initialized in the constructor. These refs will be used to store references to the DOM elements that the Popover interacts with.
* Technically `domContainer` does not need to be changed into a reference (it does not affect functionality nor result in any issues not to do it), but it makes sense to keep both dom elements as references in order to future-proof it and avoid more issues such as the findDOMNode issue. 
* The third instance where findDOMNode was used, was for `ignoreContext`. With this we're just checking the current value.


### Other information:


## Jetpack product discussion

https://github.com/Automattic/jetpack/issues/37408

## Does this pull request change what data or activity we track or use?

No.

## Testing instructions:

Testing functionality:
* With this PR checked out locally or via the Beta tester plugin, ensure that existing Popover functionality has not been changed. An example of where this Popover is used is on the Jetpack dashboard in wp-admin, - see the small info icons. Clicking them should open the Popover, and clicking them again (or elsewhere on the page) should close it.
* Note that the 'createRoot' console error is related to the RootChild component, which the Popover component uses.

Running tests:
* To actually replicate the issue, within a local development environment and with this PR checked out, replace the `popover/index.jsx` code with the existing code: https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/jetpack/_inc/client/components/popover/index.jsx
* Navigate to the Jetpack plugin folder, and from there run run the following  test - this will run all test-adminpage tests, including the popover tests: `npm run test-adminpage`. 
* Note the three failed tests - in particular the 'should not show Popover when context is not a DOM element' which mentions the findDOMNode is deprecated. The 'should handle ignoreContext correctly' test will also show this same deprecation notice, if the first two tests are commented out (or preceded by 'x', eg `xit( 'should not show Popover`...).  These three (the first two, and the ignoreContext) tests cover two of the three findDOMNode instances in the old code - I was unable to test the third but the logic is the same as in the first example.
* Then, apply the changes to the `popover/index.jsx` file from this PR and run the tests again. They should all pass.
